src/TorchSharp/DisposeScopeManager.cs

```diff
@@ -0,0 +1,317 @@
+    {
+        internal static int _disposedTensorCount;
+        internal static int _disposedCount;
+
+        static DisposeScopeManager()
+        {
+            torch.Tensor.OnTensorCreateCallback.Add(TryRegisterInDisposeScope);
+            torch.Tensor.OnTensorDisposeCallback.Add(TryUnregisterInDisposeScope);
+        }
+
+        [ThreadStatic] private static DisposeScopeManager _singleton;
+
+        public static DisposeScopeManager Singleton = (_singleton ??= new DisposeScopeManager());
+
+        internal Stack<DisposeScope> DisposeScopeStack { get; } = new();
+        internal IDisposable CurrentlyDisposing { get; set; }
+
+        public static int DisposedTensorCount => _disposedTensorCount;
+        public static int DisposedCount => _disposedCount;
+
+        public static long TotalAllocatedSize => Singleton.InnerTotalAllocatedSize;
+
+        internal static void TryUnregisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryUnregisterInDisposeScope(disposable);
+
+        internal static void TryRegisterInDisposeScope(IDisposable disposable) =>
+            Singleton.InnerTryRegisterInDisposeScope(disposable);
+
+        private long InnerTotalAllocatedSize {
+            get {
+                long size = 0;
+                foreach (var disposeScope in DisposeScopeStack) {
+                    size += disposeScope.TotalAllocatedSize;
+                }
+
+                return size;
+            }
+        }
+
+        private void InnerTryRegisterInDisposeScope(IDisposable disposable)
+        {
+            if (DisposeScopeStack.Count > 0) {
+                DisposeScopeStack.Peek().Include(disposable);
+            }
+        }
+
+        private void InnerTryUnregisterInDisposeScope(IDisposable disposable)
+        {
+            // It's us! We're disposing of it.
+            if (disposable == CurrentlyDisposing) {
+                return;
+            }
+
+            foreach (var disposeScope in DisposeScopeStack) {
+                bool wasRemoved = disposeScope.Disposables.Remove(disposable);
+                if (wasRemoved && !disposeScope.DontReportOnExternalDisposes) {
+                    // File.AppendAllText(@"c:\slask\disp.txt",
+                    //     "\n\nWho did it:"+
+                    //     disposeScope.StackTrace.ToString().Substring(0, 1000));
+                }
+            }
+        }
+
+        internal static DisposeScope NewDisposeScope()
+        {
+            lock (Singleton ??= new DisposeScopeManager()) {
+                return Singleton.InnerNewDisposeScope();
+            }
+        }
+
+        private DisposeScope InnerNewDisposeScope()
+        {
+            var disposeScope = new DisposeScope(this);
+            DisposeScopeStack.Push(disposeScope);
+            return disposeScope;
+        }
+
+        internal void RemoveDisposeScope(DisposeScope disposeScope)
+        {
+            lock (this) {
+                if (DisposeScopeStack.Count == 0) {
```

Comment:
These tests seem like something really fundamental has been going wrong. What is a scenario where this could happen?

Should they be Debug.Assert() instead?

Reply:
Absolutely, I'll change to that instead.

Reply:
When you edit the PR, please also start a new entry for version 0.95.4 in the RELEASENOTE.md file and describe the changes. This falls under 'API Changes'